packages/core/src/dialects/abstract/query-generator-typescript.ts

```diff
@@ -742,20 +762,147 @@ export class AbstractQueryGeneratorTypeScript {
         `ORDER BY ${pks}`,
-        this.#internals.addLimitAndOffset(options),
+        this.#internals.addLimitAndOffset(whereOptions),
         ')',
       ]);
     }
 
     return joinSQLFragments([
-      `DELETE FROM ${this.quoteTable(tableOrModel)}`,
-      options.where ? this.whereQuery(options.where, whereOptions) : '',
-      this.#internals.addLimitAndOffset(options),
+      `DELETE FROM ${table}`,
+      whereOptions.where ? this.whereQuery(whereOptions.where, whereOptions) : '',
+      this.#internals.addLimitAndOffset(whereOptions),
     ]);
   }
 
+  /**
+   * Returns an update query
+   *
+   * @param tableOrModel
+   * @param attrValueHash
+   * @param options
+   */
+  updateQuery(
+    tableOrModel: TableOrModel,
+    attrValueHash: Record<string, unknown>,
+    options?: UpdateQueryOptions,
+  ): QueryWithBindParams {
+    if (options) {
+      rejectInvalidOptions(
+        'updateQuery',
+        this.dialect,
+        UPDATE_QUERY_SUPPORTABLE_OPTIONS,
+        {
+          ...this.dialect.supports.update,
+          limit: this.dialect.supports.update.limit || this.dialect.supports.update.modelWithLimit,
+        },
+        options,
+      );
+    }
+
+    if (typeof attrValueHash !== 'object' || attrValueHash == null || Array.isArray(attrValueHash)) {
+      throw new Error(`Invalid value: ${NodeUtil.inspect(attrValueHash)}. Expected an object.`);
+    }
+
+    const bind = Object.create(null);
+    const valueHash = removeNullishValuesFromHash(attrValueHash, this.options.omitNull ?? false);
+    const attributeMap = new Map<string, NormalizedDataType>();
+    const modelDefinition = extractModelDefinition(tableOrModel);
+    const updateOptions: UpdateQueryOptions = {
+      ...options,
+      model: modelDefinition,
+      bindParam: options?.bindParam === undefined ? createBindParamGenerator(bind) : options.bindParam,
+    };
+
+    if (this.sequelize.options.dialectOptions.prependSearchPath || updateOptions.searchPath) {
+      // Not currently supported with search path (requires output of multiple queries)
+      updateOptions.bindParam = undefined;
+    }
+
+    if (modelDefinition) {
+      if (updateOptions.columnTypes) {
+        throw new Error('Using options.columnTypes in updateQuery requires specifying a model or model definition.');
+      }
+
+      for (const [column, attribute] of modelDefinition.physicalAttributes) {
+        attributeMap.set(attribute.columnName ?? column, attribute.type);
+      }
+    } else if (options?.columnTypes) {
+      for (const column of Object.keys(options.columnTypes)) {
+        attributeMap.set(column, this.sequelize.normalizeDataType(options.columnTypes[column]));
+      }
+    }
+
+    const updateFragment: string[] = [];
+    for (const column of Object.keys(valueHash)) {
+      const rowValue = valueHash[column];
+      if (modelDefinition?.columns.get(column)?.autoIncrement && !this.dialect.supports.autoIncrement.update) {
```

Comment:
there is a problem here: if a table is provided, then the hash is a mapping of `columnName -> value`, but if a model/modelDefinition is provided, then the hash is a mapping of `attributeName -> value`

e.g. the following two examples should work:

```ts
class Project extends Model {
  @Attribute(DataTypes.INTEGER)
  @ColumnName('user_id')
  declare userId: number;
}

await sequelize.queryGenerator.updateQuery(Project, {
  userId: 5
});

// UPDATE "projects" SET "user_id" = 5;

await sequelize.queryGenerator.updateQuery(Project.table, {
  user_id: 5
});

// UPDATE "projects" SET "user_id" = 5;
```

However, the following should throw:

```ts
await sequelize.queryGenerator.updateQuery(Project, {
  user_id: 5
});

// Error: Model "Project" does not include an attribute named "user_id".
```

But the following should of course work, because when using tables we don't know what the available attributes are:

```ts
await sequelize.queryGenerator.updateQuery(Project.table, {
  userId: 5
});

// UPDATE "projects" SET "userId" = 5;
```

There are two parts to this comment:

- if the modelDefinition is available, then the `attrValueHash` keys should be mapped to column names (you can call `getColumnName` on the model definition, it will throw if it does not exist)
- this check should use `modelDefinition.physicalAttributes` instead of `modelDefinition.columns`

There is also a bug in `ModelDefinition`: `getColumnName` should use `physicalAttributes` instead of `attributes`

Reply:
I've re-worked the model methods that called `updateQuery` and updated the tests which should allow for this to work.